torch/_functorch/_aot_autograd/autograd_cache.py

```diff
@@ -3,19 +3,79 @@
 """
 from __future__ import annotations
 
+import functools
 import logging
-from typing import TYPE_CHECKING
+import os
 
-from torch._inductor.codecache import _ident, FxGraphCachePickler
+import torch
+from torch._inductor.codecache import (
+    _ident,
+    FxGraphCachePickler,
+    get_code_hash,
+    get_inductor_root,
+)
+from torch.fx.node import Node
 
 from .schemas import AOTConfig  # noqa: F401
 
-if TYPE_CHECKING:
-    import torch
-
 log = logging.getLogger(__name__)
 
 
+class BypassAOTAutogradCache(Exception):
+    pass
+
+
+def check_node_safe(node: Node):
+    """
+    Checks that the node only uses supported operators.
+    """
+
+    def is_torch_function(target):
+        return type(target).__name__.startswith("builtin")
+
+    def is_tensor(target):
+        # Tensors always have example values in meta field
+        return "example_value" in target.meta
+
+    # I'd love to use a match statement here, but it wasn't introduced until py3.10
+    if node.op == "call_function":
+        # We support only torch.* functions for now
+        # We can probably add an allowlist of safe non-torch implementations as well
+        if not is_torch_function(node.target):
+            raise BypassAOTAutogradCache(
+                f"Unsupported call_function target {node.target}"
+            )
+    elif node.op == "call_method":
+        method_target = node.args[0]
+        if not is_tensor(method_target):
+            # We support only method calls on tensors and symints
+            raise BypassAOTAutogradCache(
+                f"Unsupported call_method target {method_target}"
+            )
+    # Cache safe
+    elif node.op in ("placeholder", "call_module", "get_attr", "output"):
+        # TODO: not all call_modules may be safe
```

Comment:
Is the approach you're planning to take to start "extremely conservative but safe", and slowly expand coverage? Or to fix correctness problems later.

(Just checking since as you pointed out, I'm not sure if call_module is always safe)

Reply:
Oops, I forgot about that: I think our assumption was that call_modules would generally be safe because we're calling torch.nn.Module. But we should be careful if there are modules that don't save state. I think just completely disallowing all call_module might be *too* conservative, right? Maybe the check is something like "if it's a torch.nn.Module" then it's safe?

Reply:
actually yep I'm pretty sure you're right:

(1) today the only call_module ops that can show up in a graph come from "built-in-nn-modules" that dynamo assumes are safe to trace. If dynamo assumes they are safely to blindly trace, then they should be safe to cache as well

(2) in the steady-state (some time in H2?) we shouldn't see these anymore, once @anijain2305 turns on "inlining builtin nn modules by default"

(3) I was also thinking about allow_in_graph + user nn modules (in another comment), but I don't think you can actually use allow_in_graph on an nn module today (only on function calls).